### DISCUSSION TITLE
urql subscription over websockets not communicating with uvicorn server

### DISCUSSION THREAD

**Discussion by markmiscavage** [urql-graphql/urql#2984] [Q&A] [answered]
I've been working on a strange issue for several days now and I believe I have the issue narrowed down, but I wanted to get the urql community feedback on what I'm seeing. 

I have a backend built with FastAPI + Strawberry running with Uvicorn. I'm able to connect, subscribe, send/receive messages to my graphql websocket endpoint there with test clients such as GraphiQL and Altair without any issues.  When I try to connect using urql and graphql-ws the websocket connection is accepted, opened, one message is sent to the server at which point the server closes the connection with error 4400 and no reason given. The requests never make it beyond Uvicorn in the stack. 

Running my own fork of Uvicorn locally with additional logging has shown me that urql is sending additional information when subscribing and if I strip that information out at the Uvicorn level the websockets work as expected. I tried both the websocket and wsproto protocol implementations with Uvicorn and both result in the exact same issue. The information that needs to be stripped out of the subscribe message being sent is `context` and `key`. Examples: 

urql subscribe message that fails:
`{'type': 'websocket.receive', 'text': '{"id":"751b595e-eb9f-4957-b65f-b15f46f3e371","type":"subscribe","payload":{"key":"-2b2a76n","query":"subscription ($userId: String!) {\\n subscribe(userId: $userId) {\\n  id\\n  key\\n  userId\\n  data\\n  timestamp\\n }\\n}","variables":{"user_id":"42"},"context":{"url":"http://localhost:8000/graphql","preferGetMethod":false,"requestPolicy":"cache-first","suspense":false,"meta":{"cacheOutcome":"miss"}}}}'}`

GraphiQL & Altair message that works:
`{'type': 'websocket.receive', 'text': '{"id":"936315c0-d47b-4958-8468-52f5f1dd8e03","type":"subscribe","payload":{"query":"subscription {\\n  subscribe(userId: \\"42\\") {\\n    id\\n    userId\\n    key\\n    data\\n    timestamp\\n  } \\n}"}}'}`

I'm currently working on stripping `context` and `key` out at the JS level to see if this fixes our issue, but I'm not even sure that's possible or reasonable. Is this an oversight and a bug in urql? Is the example urql implementation just out of date? Something else? Any thoughts would be appreciated.

**Comment by markmiscavage** [comment]
Quick update: Passing a null context to useSubscription appears to not be allowed and setting it to an empty object causes an infinite loop.

**Comment by kitten** [comment, marked as answer]
Just saying, but, Websocket transports have basically nothing to do with urql and the transports are separate packages 😅

I'd probably start asking there, but that said, I'm not even sure which GraphQL WebSockets transport you're using; is it graphql-ws?

**Reply by markmiscavage** [reply]
Yes we are using graphql-ws.  I understand the separation of the packages but it seems to me that the extra message data (`content` and `key`) are being added by urql. Most obviously seen here https://github.com/urql-graphql/urql/blob/main/packages/react-urql/src/hooks/useSubscription.test.tsx#L64-L79 where the subscription test is setting and expecting those keys. Context is part of useSubscriptionArgs and I believe key is part of caching.

**Reply by kitten** [reply]
Yea, so, basically the shape of the input to the WebSockets subscriptions client isn't necessarily what your Websockets client expects, so once you receive that and hand it over you'll have to make sure to filter it down, if the WebSockets client doesn't, unless you wish to send this info over as well

**Reply by markmiscavage** [reply]
I'm not trying to point fingers, or place blame; I'm happy to have made wrong assumptions about what I'm seeing or done something very dumb. I've tried just about everything I can think of to get urql + graphql-ws to talk to my FastAPI + Strawberry + Uvicorn server and they won't do it despite other implementations working.

**Reply by kitten** [reply]
Well, if you want to match the payload of GraphiQL exactly and you're using `graphql-ws`, then I assume either it doesn't filter, this specific version doesn't filter, or it stopped filtering, since I'm seeing that it doesn't match the snippet in our docs.
However, you should always be able to set up the following, as per our [example](https://formidable.com/open-source/urql/docs/advanced/subscriptions/) with the added remapping:

```ts
import { createClient, defaultExchanges, subscriptionExchange } from 'urql';
import { createClient as createWSClient } from 'graphql-ws';

const wsClient = createWSClient({
  url: 'ws://localhost/graphql',
});

const client = createClient({
  url: '/graphql',
  exchanges: [
    ...defaultExchanges,
    subscriptionExchange({
      forwardSubscription: (operation) => ({
        subscribe: (sink) => ({
          unsubscribe: wsClient.subscribe({ query: operation.query, variables: operation.variables }, sink),
        }),
      }),
    }),
  ],
});
```

**Reply by markmiscavage** [reply]
That looks like its working now. Thank you for the help!
